### PR TITLE
Fixes compression method override of zero-length files

### LINKED ISSUE #30317: Zip package becomes corrupted (CRC error)
Refer to https://github.com/OfficeDev/Open-XML-SDK/issues/506

This is an issue I first discovered in dealing with OpenXml and it seems the root cause is a bug in `System.IO.Packaging`.

**Description**

Using `System.IO.Packaging` to purely open and save a file (without any manipulation) causes a corrupt zip package. 7-Zip says there is a CRC error:
image

This subsequently means that Excel 2013 complains that the `xlsx` file is corrupt.

Note that the file contains custom property parts, some of which are of 0 lengths. I don't have control of those parts.

This problem occurs *only* when using .NET Core, *not with .NET Framework*.

**Information**

- .NET Target: .NET Core 2.0
- System.IO.Packaging Version: v.4.5.0

**Repro**

```csharp
using System;
using System.IO;
using System.IO.Packaging;

namespace OpenXMLCustomPropertyPackagingProblemCore
{
    internal class Program
    {
        private static void Main(string[] args)
        {
            File.Copy("test.xlsx", "file_to_open.xlsx", true);
            using (var package = Package.Open("file_to_open.xlsx", FileMode.Open))
            {
            }

            Console.WriteLine("Done.");
            Console.ReadKey(false);
        }
    }
}
``` 
Files 
- [test.xlsx](https://github.com/dotnet/corefx/files/2094171/test.xlsx)
- [file_to_open.xlsx](https://github.com/dotnet/corefx/files/2094173/file_to_open.xlsx) (produced by running the code)

MD5 hashes:
```
94db2b0108899c032526987091eaea6f *file_to_open.xlsx
ac39d841fc15384c252c4870063f9b54 *test.xlsx
```

### PR DESCRIPTION
**Pull request by jimdemis**
Fixes #30317. When a zero-length file is stored using compression method, ZipArchive still overrides the method to “not compressed”, and ignore compressed bytes that are still present, rendering the zip file unusable.

According to [zip format specifications 4.3.8]( https://pkware.cachefly.net/webdocs/casestudies/APPNOTE.TXT):
` Zero-byte files, directories, and other file types that contain no content **MUST NOT** include file data.
`

Reproduction:
1. Zip file contains zero-byte file marked as `Deflate 0x08`, uncompressed size `0`, compressed `2` bytes and compressed `file data = \u0003 (ETX)`
2. Calling ZipArchive constructor and disposing it will trigger archive rewriting
3. While rewriting, ZipArchive does not expect zero-byte file with `uncompressedSize==0` to have `file data`, so it changes the compression method to `Stored 0x00`, but keeps the `file data`
4. The `file data` left inside the archive, will render it unusable.

Possible solutions:
1. ~~Ignore the content and continue (that's what I did now)~~ (not good, will save bad data if the zip is malformed)
2. Remove the `file data` 
3. Throw `InvalidDataException`

Proposed solution:
According to zip specs, ignore the `file data` (will not be saved, thus removed) and still override the compression method to `Stored 0x00`

/cc @buyaa-n, @ahsonkhan, @ViktorHofer

### CODE REVIEW

**Comment by jimdemis** [comment]
ready to review /cc @karelz @danmosemsft @ericstj @buyaa-n, @ahsonkhan, @ViktorHofer

File: src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs [hunk 1]
```diff
@@ -912,7 +912,12 @@ private void WriteLocalFileHeaderAndDataIfNeeded()
                 {
                     // we know the sizes at this point, so just go ahead and write the headers
                     if (_uncompressedSize == 0)
+                    {
+                        // according to ZIP specs, zero-byte files MUST NOT include file data
+                        _compressedSize = 0;
                         CompressionMethod = CompressionMethodValues.Stored;
+                    }
+                        
                     WriteLocalFileHeader(isEmptyFile: false);
```
Comment (stephentoub) [inline src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs]: How does this `isEmptyFile` argument relate?  Should this be true? And if yes, does `_compressedSize` need to be change to 0?

File: src/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs [hunk 2]
```diff
@@ -241,6 +241,74 @@ public static void ZipWithLargeSparseFile()
...
+            0x2F, 0x63, 0x75, 0x73, 0x74, 0x6F, 0x6D, 0x50, 0x72, 0x6F, 0x70, 0x65, 0x72, 0x74, 0x79, 0x32,
+            0x2E, 0x62, 0x69, 0x6E, 0xBA, 0xAD, 0x03, 0x00, 0x50, 0x4B, 0x01, 0x02, 0x14, 0x00, 0x14, 0x00,
+            0x06, 0x00, 0x08, 0x00, 0x00, 0x00, 0x21, 0x00, 0x00, 0x00, 0x00, 0x00, 0x04, 0x00, 0x00, 0x00,
+            0x00, 0x00, 0x00, 0x00, 0x16, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+            0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x78, 0x6C, 0x2F, 0x63, 0x75, 0x73, 0x74, 0x6F, 0x6D, 0x50,
+            0x72, 0x6F, 0x70, 0x65, 0x72, 0x74, 0x79, 0x32, 0x2E, 0x62, 0x69, 0x6E, 0x50, 0x4B, 0x05, 0x06,
+            0x00, 0x00, 0x00, 0x00, 0x01, 0x00, 0x01, 0x00, 0x44, 0x00, 0x00, 0x00, 0x38, 0x00, 0x00, 0x00,
+            0x00, 0x00
+        };
+        internal static readonly Collections.Generic.List<byte[]> issue30317 = new Collections.Generic.List<byte[]>(){
```
Comment (stephentoub) [inline src/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs]: `issue30317` => s_emptyFiles

Comment (stephentoub) [reply src/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs]: This should be made public.

File: src/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs [hunk 3]
```diff
@@ -241,6 +241,74 @@ public static void ZipWithLargeSparseFile()
...
+        /// 1. EmptyFileCompressedWithEOT has 
+        /// Deflate 0x08, _uncompressedSize 0, _compressedSize 2, compressed data: 0x0300 (\u0003 ETX)
+        /// 2. EmptyFileCompressedWrongSize has 
+        /// Deflate 0x08, _uncompressedSize 0, _compressedSize 4, compressed data: 0xBAAD0300 (just bad data)
+        /// ZipArchive is expected to change compression method to Stored (0x00) and ignore "bad" compressed size
+        /// </summary>
+        [Fact]
+        public static void ReadArchive_WithEmptyDeflatedFile()
+        {
+            foreach (var fileBytes in issue30317)
```
Comment (stephentoub) [inline src/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs]: It'd be better to make this a `[Theory(nameof(EmptyFiles))]`, accepting the `byte[] fileBytes` to use.

File: src/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs [hunk 4]
```diff
@@ -241,6 +241,74 @@ public static void ZipWithLargeSparseFile()
...
+            {
+                using (var testStream = new MemoryStream(fileBytes))
+                {
+                    var tmpfilename = "xl/customProperty2.bin";
+                    var tmppathname = Path.GetTempPath();
+                    var tmpfilepath = tmppathname + tmpfilename;
```
Comment (stephentoub) [inline src/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs]: Please change these `var`s to `string`s.

File: src/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs [hunk 5]
```diff
@@ -241,6 +241,74 @@ public static void ZipWithLargeSparseFile()
...
+                    // open archive with zero-length file that is compressed (Deflate = 0x8)
+                    using (var zip = new ZipArchive(testStream, ZipArchiveMode.Update, true))
```
Comment (stephentoub) [inline src/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs]: Nit: please name the `true` argument so it's easier to understand at the call site what it's specifying.

File: src/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs (hunk 4 above)
Comment (stephentoub) [inline src/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs]: This should use `string tmpfilepath = GetTestFilePath();`, using the helper from the base test class.  That will ensure it's in a good temporary location with a good name, and that the file will be cleaned up after the test ends.  You can use `TestDirectory` to know what directory it's in.

Comment (stephentoub) [reply src/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs]: Oh, I see, you're not generating the file name, but rather it's already encoded into the original zip.  In that case, just use `TestDirectory` instead of `Path.GetTempPath()`.

Comment (danmosemsft) [inline src/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs]: Path.Combine?

File: src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs (hunk 1 above)
Comment (jimdemis) [reply src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs]: My bad thanks for spotting it out! 
`_compressedSize` needs to be changed because it will be used later on at the central directory, I added a comment.

File: src/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs [hunk 6]
```diff
@@ -241,6 +241,47 @@ public static void ZipWithLargeSparseFile()
...
+
+                byte[] fileContent = testStream.ToArray();
+
+                // compression method should stay the same (Deflate = 0x8)
+                Assert.Equal(8, fileContent[8]);
+
+                // extract and check the file. should stay empty.
+                using (var zip = new ZipArchive(testStream, ZipArchiveMode.Update))
+                {
+                    zip.ExtractToDirectory(tmppathname, overwriteFiles: true);
```
Comment (ericstj) [inline src/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs]: Don't leak temp files in the tests.  There is no reason you need to test with ExtractToDirectory, just enumerate entries, assert that lengths are zero, then open the streams and ensure that lengths are zero.

File: src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs [hunk 7]
```diff
@@ -804,7 +804,6 @@ private bool WriteLocalFileHeader(bool isEmptyFile)
                 CompressionMethod = CompressionMethodValues.Stored;
                 compressedSizeTruncated = 0;
                 uncompressedSizeTruncated = 0;
-                Debug.Assert(_compressedSize == 0);
```
Comment (buyaa-n) [inline src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs]: Why do you need to remove it? seems you still setting this to 0 if it's empty

File: src/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs [hunk 8]
```diff
@@ -241,6 +243,72 @@ public static void ZipWithLargeSparseFile()
...
+        /// Deflate 0x08, _uncompressedSize 0, _compressedSize 4, compressed data: 0xBAAD0300 (just bad data)
+        /// ZipArchive is expected to change compression method to Stored (0x00) and ignore "bad" compressed size
+        /// </summary>
+        [Theory]
+        [MemberData(nameof(EmptyFiles))]
+        public void ReadArchive_WithEmptyDeflatedFile(byte[] fileBytes)
+        {
+            using (var testStream = new MemoryStream(fileBytes))
+            {
+                string tmpfilename = "xl/customProperty2.bin";
```
Comment (buyaa-n) [inline src/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs]: seems this is no used anymore

File: src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs (hunk 7 above)
Comment (jimdemis) [reply src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs]: Because I was setting it to 0 after calling this method. Changed it back now.

File: src/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs (hunk 8 above)
Comment (jimdemis) [reply src/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs]: still used once in `zip.GetEntry(tmpfilename);`

Comment (jimdemis) [reply src/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs]: shoud i remove it and use `zip.GetEntry("xl/customProperty2.bin")` ?

Comment (buyaa-n) [reply src/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs]: didn't noticed that, not needed, thanks

File: src/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs [hunk 9]
```diff
@@ -16,7 +17,8 @@ public class zip_InvalidParametersAndStrangeFiles : ZipFileTestBase
             {
                 Assert.Throws<TException>(() =>
                 {
-                    using (ZipArchive archive = constructor()) { }
+                    using (ZipArchive archive = constructor())
+                    { }
```
Comment (stephentoub) [inline src/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs]: This can be reverted.  It's just a whitespace change.

File: src/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs [hunk 10]
```diff
@@ -241,6 +243,77 @@ public static void ZipWithLargeSparseFile()
...
+            0x00, 0x00, 0x02, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x16, 0x00, 0x00, 0x00, 0x78, 0x6C,
+            0x2F, 0x63, 0x75, 0x73, 0x74, 0x6F, 0x6D, 0x50, 0x72, 0x6F, 0x70, 0x65, 0x72, 0x74, 0x79, 0x32,
+            0x2E, 0x62, 0x69, 0x6E, 0x03, 0x00, 0x50, 0x4B, 0x01, 0x02, 0x14, 0x00, 0x14, 0x00, 0x06, 0x00,
+            0x08, 0x00, 0x00, 0x00, 0x21, 0x00, 0x00, 0x00, 0x00, 0x00, 0x02, 0x00, 0x00, 0x00, 0x00, 0x00,
+            0x00, 0x00, 0x16, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+            0x00, 0x00, 0x00, 0x00, 0x78, 0x6C, 0x2F, 0x63, 0x75, 0x73, 0x74, 0x6F, 0x6D, 0x50, 0x72, 0x6F,
+            0x70, 0x65, 0x72, 0x74, 0x79, 0x32, 0x2E, 0x62, 0x69, 0x6E, 0x50, 0x4B, 0x05, 0x06, 0x00, 0x00,
+            0x00, 0x00, 0x01, 0x00, 0x01, 0x00, 0x44, 0x00, 0x00, 0x00, 0x36, 0x00, 0x00, 0x00, 0x00, 0x00
+        };
+        internal static readonly byte[] EmptyFileCompressedWrongSize =
```
Comment (stephentoub) [inline src/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs]: Nit: these two fields can be private and use s_camelCasedNaming.  The one below it that uses them is good as-is with being public and PascalCased Naming.

File: src/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs [hunk 11]
```diff
@@ -241,6 +243,77 @@ public static void ZipWithLargeSparseFile()
...
+        /// Deflate 0x08, _uncompressedSize 0, _compressedSize 4, compressed data: 0xBAAD0300 (just bad data)
+        /// ZipArchive is expected to change compression method to Stored (0x00) and ignore "bad" compressed size
+        /// </summary>
+        [Theory]
+        [MemberData(nameof(EmptyFiles))]
+        public void ReadArchive_WithEmptyDeflatedFile(byte[] fileBytes)
+        {
+            using (var testStream = new MemoryStream(fileBytes))
+            {
+                string tmpfilename = "xl/customProperty2.bin";
```
Comment (stephentoub) [inline src/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs]: Nit: `const string ExpectedFileName`

**Comment by igitur** [comment]
Thanks @jimdemis
